# 第3章 章末演習問題 (全24問)

教科書「パターン認識と機械学習 (PRML)」第3章「線形回帰モデル (Linear Models for Regression)」のすべての Exercises (3.1 〜 3.24) を網羅しています。
学習者のモチベーションを維持しながら自力で論理的展開と思考を追体験できるよう、**論理ステップの提示＋穴埋め・選択形式**および**Pythonによる数値検証コード**で構成されています。

---
## 目次
- [3.1 tanh とシグモイド関数の関係](#Exercise-3.1)
- [3.2 最小二乗直交射影行列の性質](#Exercise-3.2)
- [3.3 重み付き二乗和誤差と2つの解釈](#Exercise-3.3)
- [3.4 入力ノイズとWeight Decay正則化の等価性](#Exercise-3.4)
- [3.5 ラグランジュ未定乗数法と正則化制約](#Exercise-3.5)
- [3.6 多変量目的変数の最尤推定](#Exercise-3.6)
- [3.7 平方完成によるパラメータ事後分布の導出](#Exercise-3.7)
- [3.8 逐次ベイズ更新の平方完成](#Exercise-3.8)
- [3.9 線形ガウス公式による事後分布導出](#Exercise-3.9)
- [3.10 周辺化による予測分布の導出](#Exercise-3.10)
- [3.11 Sherman-Morrison公式と予測分散の単調減少](#Exercise-3.11)
- [3.12 未知の平均と精度に対する正規ガンマ共役事前分布](#Exercise-3.12)
- [3.13 正規ガンマモデルのStudent's t予測分布](#Exercise-3.13)
- [3.14 正規直交基底と等価カーネルの総和制約](#Exercise-3.14)
- [3.15 エビデンス停留点における $2E(\mathbf{m}_N) = N$ の証明](#Exercise-3.15)
- [3.16 線形ガウス公式による対数エビデンスの直接導出](#Exercise-3.16)
- [3.17 エビデンス関数の指数部の変形](#Exercise-3.17)
- [3.18 平方完成による誤差関数の変形](#Exercise-3.18)
- [3.19 ガウス積分による対数周辺尤度の導出](#Exercise-3.19)
- [3.20 対数エビデンスの $\alpha$ 微分と再推定式の導出](#Exercise-3.20)
- [3.21 行列式微分恒等式を用いた $\alpha$ の再推定式導出](#Exercise-3.21)
- [3.22 対数エビデンスの $\beta$ 微分と再推定式の導出](#Exercise-3.22)
- [3.23 正規ガンマモデルのエビデンスの直接積分導出](#Exercise-3.23)
- [3.24 ベイズの定理によるエビデンスの簡潔な導出](#Exercise-3.24)
---

In [1]:
import sys, os
sys.path.append(os.path.abspath('../'))
import numpy as np
import scipy.stats as stats
from common.regression_utils import PolynomialBasis, GaussianBasis, SigmoidalBasis, LinearRegression, RidgeRegression, BayesianLinearRegression
print("Setup completed successfully.")

Setup completed successfully.


## Exercise 3.1
**問題**: 双曲線正接関数 $\tanh(a) = \frac{e^a - e^{-a}}{e^a + e^{-a}}$ とロジスティックシグモイド関数 $\sigma(a) = \frac{1}{1 + e^{-a}}$ の間に
$$ \tanh(a) = 2\sigma(2a) - 1 $$
の関係が成り立つことを示せ。
また、シグモイド基底の線形結合
$$ y(x, \mathbf{w}) = w_0 + \sum_{j=1}^M w_j \sigma\left(\frac{x - \mu_j}{s}\right) $$
が $\tanh$ 基底の線形結合
$$ y(x, \mathbf{u}) = u_0 + \sum_{j=1}^M u_j \tanh\left(\frac{x - \mu_j}{2s}\right) $$
と等価であることを示し、新パラメータ $\{u_0, u_1, \dots, u_M\}$ と元パラメータ $\{w_0, w_1, \dots, w_M\}$ の変換関係を求めよ。

### 証明・導出の論理ステップ
1. $\sigma(2a) = \frac{1}{1 + e^{-2a}} = \frac{e^a}{e^a + e^{-a}}$。
2. $2\sigma(2a) - 1 = \frac{2e^a - (e^a + e^{-a})}{e^a + e^{-a}} = \frac{e^a - e^{-a}}{e^a + e^{-a}} = \tanh(a)$。
3. したがって $\sigma(a) = \frac{1}{2}\{1 + \tanh(a/2)\}$。
4. これを $y(x, \mathbf{w})$ に代入すると：
   $$ y(x, \mathbf{w}) = w_0 + \sum_{j=1}^M w_j \frac{1}{2}\left\{1 + \tanh\left(\frac{x - \mu_j}{2s}\right)\right\} = \left(w_0 + \frac{1}{2}\sum_{j=1}^M w_j\right) + \sum_{j=1}^M \frac{w_j}{2} \tanh\left(\frac{x - \mu_j}{2s}\right) $$
5. したがって、
   - $u_0 = [ \text{穴埋め 1} ]$
   - $u_j = [ \text{穴埋め 2} ]$ ($j \ge 1$)

In [2]:
# Exercise 3.1 数値検証
# YOUR CODE HERE: u0 と u_j の関係を実装して y_sigma と y_tanh が完全一致することを確認
w0 = 1.5
w = np.array([0.8, -1.2, 2.0])
mu = np.array([-0.5, 0.0, 0.5])
s = 0.3
x_test = np.linspace(-1, 1, 100)

def sigmoid(a):
    return 1.0 / (1.0 + np.exp(-a))

# シグモイド基底での出力
y_sigma = w0 + np.sum([w[j] * sigmoid((x_test - mu[j])/s) for j in range(len(w))], axis=0)

# パラメータ変換の穴埋め
# u0 = ...
# u = ...
u0 = w0 + 0.5 * np.sum(w)  # YOUR CODE HERE
u = 0.5 * w               # YOUR CODE HERE

# tanh 基底での出力 (スケールは 2s)
y_tanh = u0 + np.sum([u[j] * np.tanh((x_test - mu[j])/(2*s)) for j in range(len(u))], axis=0)

diff = np.max(np.abs(y_sigma - y_tanh))
print(f"Max difference: {diff:.2e}")
assert diff < 1e-12, "Test failed!"
print("Exercise 3.1 passed successfully!")

Max difference: 8.88e-16
Exercise 3.1 passed successfully!


## Exercise 3.2
**問題**: 行列 $\mathbf{P} = \mathbf{\Phi}(\mathbf{\Phi}^T \mathbf{\Phi})^{-1} \mathbf{\Phi}^T$ が任意のベクトル $\mathbf{v}$ を $\mathbf{\Phi}$ の列が張る部分空間 $\mathcal{S}$ に直交射影することを示せ。また、これを用いて最小二乗解 (3.15) が目標値ベクトル $\mathbf{t}$ の部分空間 $\mathcal{S}$ への直交射影に対応することを示せ（PRML Figure 3.2）。

### 証明の論理ステップ
1. **部分空間に含まれること**: $\mathbf{P}\mathbf{v} = \mathbf{\Phi} [(\mathbf{\Phi}^T \mathbf{\Phi})^{-1} \mathbf{\Phi}^T \mathbf{v}]$ であり、大括弧内は $M$ 次元ベクトルであるから、$\mathbf{P}\mathbf{v}$ は明らかに $\mathbf{\Phi}$ の列ベクトルの線形結合、すなわち部分空間 $\mathcal{S}$ に属する。
2. **射影行列の性質 (冪等性)**: $\mathbf{P}^2 = \mathbf{\Phi}(\mathbf{\Phi}^T \mathbf{\Phi})^{-1} \mathbf{\Phi}^T \mathbf{\Phi}(\mathbf{\Phi}^T \mathbf{\Phi})^{-1} \mathbf{\Phi}^T = [ \text{穴埋め 1: P} ]$。
3. **残差の直交性**: 任意の $\mathbf{v}$ に対し、残差ベクトル $\mathbf{v} - \mathbf{P}\mathbf{v}$ と $\mathbf{\Phi}$ の任意の列ベクトルとの内積を計算すると：
   $$ \mathbf{\Phi}^T (\mathbf{v} - \mathbf{P}\mathbf{v}) = \mathbf{\Phi}^T \mathbf{v} - \mathbf{\Phi}^T \mathbf{\Phi}(\mathbf{\Phi}^T \mathbf{\Phi})^{-1} \mathbf{\Phi}^T \mathbf{v} = [ \text{穴埋め 2: 0} ] $$
   ゆえに残差は部分空間 $\mathcal{S}$ のすべての基底ベクトルと直交する。
4. **最小二乗解との一致**: 予測値ベクトル $\mathbf{y} = \mathbf{\Phi} \mathbf{w}_{\mathrm{ML}} = \mathbf{\Phi} (\mathbf{\Phi}^T \mathbf{\Phi})^{-1} \mathbf{\Phi}^T \mathbf{t} = \mathbf{P}\mathbf{t}$ より、最小二乗解は目標値 $\mathbf{t}$ の部分空間 $\mathcal{S}$ への直交射影である。

In [3]:
# Exercise 3.2 数値検証
np.random.seed(0)
N, M = 10, 4
Phi = np.random.randn(N, M)
v = np.random.randn(N)

# YOUR CODE HERE: P の計算と直交性検証
P = Phi @ np.linalg.inv(Phi.T @ Phi) @ Phi.T
# 冪等性 P^2 == P
assert np.allclose(P @ P, P), "Idempotence failed"
# 対称性 P^T == P
assert np.allclose(P.T, P), "Symmetry failed"
# 残差の直交性 Phi^T @ (v - P @ v) == 0
assert np.allclose(Phi.T @ (v - P @ v), np.zeros(M)), "Orthogonality failed"
print("Exercise 3.2 passed successfully!")

Exercise 3.2 passed successfully!


## Exercise 3.3
**問題**: 各データ点 $t_n$ に重み係数 $r_n > 0$ が付与された重み付き二乗和誤差
$$ E_D(\mathbf{w}) = \frac{1}{2} \sum_{n=1}^N r_n \{t_n - \mathbf{w}^T \boldsymbol{\phi}(\mathbf{x}_n)\}^2 $$
を最小化する解 $\mathbf{w}^*$ を求めよ。また、この誤差関数の2つの解釈（(i) データ依存ノイズ分散、(ii) 重複データ点）を述べよ。

### 導出と解釈の論理ステップ
1. 行列形式で表すと、$E_D(\mathbf{w}) = \frac{1}{2} (\mathbf{t} - \mathbf{\Phi}\mathbf{w})^T \mathbf{R} (\mathbf{t} - \mathbf{\Phi}\mathbf{w})$。ここで $\mathbf{R} = \mathrm{diag}(r_1, \dots, r_N)$。
2. $\mathbf{w}$ で微分して 0 と置くと：
   $$ \nabla E_D(\mathbf{w}) = -\mathbf{\Phi}^T \mathbf{R} (\mathbf{t} - \mathbf{\Phi}\mathbf{w}) = \mathbf{0} \implies \mathbf{w}^* = (\mathbf{\Phi}^T \mathbf{R} \mathbf{\Phi})^{-1} \mathbf{\Phi}^T \mathbf{R} \mathbf{t} $$
3. **解釈 (i) データ依存ノイズ**:
   各データ点ごとにノイズの精度（逆分散）が異なる場合 $p(t_n|\mathbf{x}_n, \mathbf{w}) = \mathcal{N}(t_n | \mathbf{w}^T\boldsymbol{\phi}_n, \beta_n^{-1})$、対数尤度関数は $-\frac{1}{2}\sum \beta_n (t_n - \mathbf{w}^T\boldsymbol{\phi}_n)^2 + \text{const}$ となり、$r_n = \beta_n$ と見なせる。ノイズが小さい高精度な観測点ほど大きな重みが与えられる。
4. **解釈 (ii) 重複データ点**:
   整数 $r_n$ に対し、全く同一のデータ点 $(\mathbf{x}_n, t_n)$ がデータセット中に $r_n$ 回重複して観測された場合と数学的に完全に同値である。

In [4]:
# Exercise 3.3 数値検証
np.random.seed(42)
N, M = 20, 3
Phi = np.random.randn(N, M)
t = np.random.randn(N)
r = np.random.uniform(0.5, 2.0, N)
R = np.diag(r)

# 重み付き最小二乗解
w_star = np.linalg.solve(Phi.T @ R @ Phi, Phi.T @ R @ t)

# 数値勾配との一致確認
def weighted_loss(w):
    return 0.5 * np.sum(r * (t - Phi @ w)**2)

grad_num = np.zeros(M)
eps = 1e-6
for i in range(M):
    w_p = w_star.copy(); w_p[i] += eps
    w_m = w_star.copy(); w_m[i] -= eps
    grad_num[i] = (weighted_loss(w_p) - weighted_loss(w_m)) / (2 * eps)

assert np.allclose(grad_num, np.zeros(M), atol=1e-5), "Gradient at w_star is not zero"
print("Exercise 3.3 passed successfully!")

Exercise 3.3 passed successfully!


## Exercise 3.4
**問題**: 線形モデル $y(\mathbf{x}, \mathbf{w}) = w_0 + \sum_{i=1}^D w_i x_i$ において、入力変数 $x_i$ に平均 0、分散 $\sigma^2$ の互いに独立なガウスノイズ $\epsilon_i$ が加わるものとする。
$\mathbb{E}[\epsilon_i] = 0, \mathbb{E}[\epsilon_i \epsilon_j] = \delta_{ij}\sigma^2$ を用いて、ノイズ分布に関して期待値をとった二乗誤差が、**バイアスパラメータ $w_0$ を除外した L2 正則化（Weight Decay）二乗誤差**と等価になることを証明せよ。

### 証明の論理ステップ
1. ノイズの加わった入力は $\tilde{x}_{ni} = x_{ni} + \epsilon_{ni}$。
2. 予測値は $\tilde{y}_n = w_0 + \sum_i w_i (x_{ni} + \epsilon_{ni}) = y(\mathbf{x}_n, \mathbf{w}) + \mathbf{w}^T \boldsymbol{\epsilon}_n$ （ただし $\mathbf{w} = (w_1, \dots, w_D)^T$）。
3. 誤差は $\tilde{y}_n - t_n = (y(\mathbf{x}_n, \mathbf{w}) - t_n) + \mathbf{w}^T \boldsymbol{\epsilon}_n$。
4. 二乗してノイズに関する期待値をとると：
   $$ \mathbb{E}_{\boldsymbol{\epsilon}} [(\tilde{y}_n - t_n)^2] = (y(\mathbf{x}_n, \mathbf{w}) - t_n)^2 + 2(y(\mathbf{x}_n, \mathbf{w}) - t_n) \mathbf{w}^T \underbrace{\mathbb{E}[\boldsymbol{\epsilon}_n]}_{\mathbf{0}} + \mathbb{E}\left[ \left(\sum_{i=1}^D w_i \epsilon_{ni}\right)^2 \right] $$
5. 交叉項の期待値は 0 となり、第3項は
   $$ \sum_{i, j} w_i w_j \mathbb{E}[\epsilon_{ni}\epsilon_{nj}] = \sigma^2 \sum_{i=1}^D w_i^2 = \sigma^2 \|\mathbf{w}\|^2 $$
6. 全 $N$ データにわたって和をとると：
   $$ \mathbb{E}_{\boldsymbol{\epsilon}}[E_D] = \frac{1}{2} \sum_{n=1}^N \{y(\mathbf{x}_n, \mathbf{w}) - t_n\}^2 + \frac{N\sigma^2}{2} \sum_{i=1}^D w_i^2 $$
   これは正則化係数 $\lambda = N\sigma^2$ でバイアス項 $w_0$ を含まない正則化項と完全に一致する！

In [5]:
# Exercise 3.4 数値検証 (モンテカルロ法による確認)
np.random.seed(42)
N, D = 50, 3
sigma = 0.2
X = np.random.randn(N, D)
w_true = np.array([1.0, -0.5, 0.8])
w0_true = 0.4
t = w0_true + X @ w_true + np.random.randn(N) * 0.1

w_eval = np.array([0.7, -0.3, 0.5])
w0_eval = 0.2

# 解析的な期待値
err_clean = 0.5 * np.sum((w0_eval + X @ w_eval - t)**2)
reg_term = 0.5 * N * (sigma**2) * np.sum(w_eval**2)
expected_loss_analytic = err_clean + reg_term

# モンテカルロ平均 (10000サンプルのノイズ平均)
n_mc = 10000
mc_losses = []
for _ in range(n_mc):
    noise = np.random.normal(0, sigma, size=X.shape)
    y_noisy = w0_eval + (X + noise) @ w_eval
    mc_losses.append(0.5 * np.sum((y_noisy - t)**2))
expected_loss_mc = np.mean(mc_losses)

print(f"Analytic Expected Loss: {expected_loss_analytic:.4f}")
print(f"Monte Carlo Mean Loss:  {expected_loss_mc:.4f}")
assert np.abs(expected_loss_analytic - expected_loss_mc) < 0.05, "Mismatch between analytic and MC"
print("Exercise 3.4 passed successfully!")

Analytic Expected Loss: 6.5738
Monte Carlo Mean Loss:  6.5866
Exercise 3.4 passed successfully!


## Exercise 3.5
**問題**: ラグランジュの未定乗数法を用いて、正則化誤差関数 (3.29)
$$ \tilde{E}(\mathbf{w}) = \frac{1}{2} \sum_{n=1}^N \{t_n - \mathbf{w}^T \boldsymbol{\phi}(\mathbf{x}_n)\}^2 + \frac{\lambda}{2} \sum_{j=1}^M |w_j|^q $$
の最小化が、制約条件 $\sum_{j=1}^M |w_j|^q \le \eta$ の下での非正則化二乗和誤差の最小化と等価であることを示せ。またパラメータ $\eta$ と $\lambda$ の関係を論ぜよ。

### 証明の論理ステップ
1. 制約付き最適化問題：
   $$ \min_{\mathbf{w}} E_D(\mathbf{w}) \quad \text{subject to} \quad \sum_{j=1}^M |w_j|^q \le \eta $$
2. カルーシュ・クーン・タッカー (KKT) 条件に基づき、ラグランジュ関数を定義する：
   $$ \mathcal{L}(\mathbf{w}, \lambda) = E_D(\mathbf{w}) + \frac{\lambda}{2} \left( \sum_{j=1}^M |w_j|^q - \eta \right) $$
3. 制約がアクティブ（等号成立）であるとき、$\mathcal{L}$ の $\mathbf{w}$ に関する停留値は $\tilde{E}(\mathbf{w})$ の最小化と定数項 $-\frac{\lambda\eta}{2}$ を除いて完全に一致する。
4. $\lambda$ と $\eta$ の関係：$\lambda$ は制約の「厳しさ」を制御するラグランジュ乗数であり、$\lambda$ を大きくするほど許容される領域の半径 $\eta$ は**単調減少**する。

In [6]:
# Exercise 3.5 確認コード
# lambda を増加させたとき、重みの L_q ノルムが単調減少することを確認 (q=2)
Phi_toy = np.random.randn(20, 5)
t_toy = np.random.randn(20)

lambdas = np.logspace(-2, 3, 20)
norms = []
for lam in lambdas:
    w_lam = np.linalg.solve(lam * np.eye(5) + Phi_toy.T @ Phi_toy, Phi_toy.T @ t_toy)
    norms.append(np.sum(w_lam**2))

# 単調減少の確認
diffs = np.diff(norms)
assert np.all(diffs <= 1e-8), "L2 norm is not monotonically decreasing with lambda"
print("Exercise 3.5 passed successfully!")

Exercise 3.5 passed successfully!


## Exercise 3.6
**問題**: 多変量目的変数 $\mathbf{t} \in \mathbb{R}^K$ に対するガウス線形基底関数回帰モデル
$$ p(\mathbf{t} | \mathbf{W}, \mathbf{\Sigma}) = \mathcal{N}(\mathbf{t} | \mathbf{W}^T \boldsymbol{\phi}(\mathbf{x}), \mathbf{\Sigma}) $$
において、パラメータ行列 $\mathbf{W} \in \mathbb{R}^{M \times K}$ の最尤推定量 $\mathbf{W}_{\mathrm{ML}}$ の各列が共分散行列 $\mathbf{\Sigma}$ に依存せず、スカラーの場合の解 (3.15) と一致することを示せ。また共分散行列の最尤推定量 $\mathbf{\Sigma}_{\mathrm{ML}}$ を求めよ。

### 証明の論理ステップ
1. 対数尤度関数：
   $$ \ln p(\mathbf{T} | \mathbf{W}, \mathbf{\Sigma}) = -\frac{NK}{2}\ln(2\pi) - \frac{N}{2}\ln|\mathbf{\Sigma}| - \frac{1}{2}\sum_{n=1}^N (\mathbf{t}_n - \mathbf{W}^T\boldsymbol{\phi}_n)^T \mathbf{\Sigma}^{-1} (\mathbf{t}_n - \mathbf{W}^T\boldsymbol{\phi}_n) $$
2. 二次形式のトレース表現：
   $$ \sum_{n=1}^N (\mathbf{t}_n - \mathbf{W}^T\boldsymbol{\phi}_n)^T \mathbf{\Sigma}^{-1} (\mathbf{t}_n - \mathbf{W}^T\boldsymbol{\phi}_n) = \mathrm{Tr}\left[ \mathbf{\Sigma}^{-1} (\mathbf{T} - \mathbf{\Phi}\mathbf{W})^T (\mathbf{T} - \mathbf{\Phi}\mathbf{W}) \right] $$
3. $\mathbf{W}$ に関する微分：
   $$ \frac{\partial}{\partial \mathbf{W}} \ln p = \mathbf{\Phi}^T (\mathbf{T} - \mathbf{\Phi}\mathbf{W}) \mathbf{\Sigma}^{-1} = \mathbf{0} $$
   両辺に右から $\mathbf{\Sigma}$ を掛けると $\mathbf{\Sigma}$ が消去され、
   $$ \mathbf{\Phi}^T \mathbf{\Phi} \mathbf{W} = \mathbf{\Phi}^T \mathbf{T} \implies \mathbf{W}_{\mathrm{ML}} = (\mathbf{\Phi}^T \mathbf{\Phi})^{-1} \mathbf{\Phi}^T \mathbf{T} $$
4. $\mathbf{\Sigma}$ に関する微分より：
   $$ \mathbf{\Sigma}_{\mathrm{ML}} = \frac{1}{N} \sum_{n=1}^N (\mathbf{t}_n - \mathbf{W}_{\mathrm{ML}}^T\boldsymbol{\phi}_n)(\mathbf{t}_n - \mathbf{W}_{\mathrm{ML}}^T\boldsymbol{\phi}_n)^T $$

In [7]:
# Exercise 3.6 数値検証
np.random.seed(42)
N, M, K = 30, 4, 2
Phi = np.random.randn(N, M)
T = np.random.randn(N, K)

W_ml = np.linalg.pinv(Phi) @ T
Sigma_ml = (1.0 / N) * (T - Phi @ W_ml).T @ (T - Phi @ W_ml)

assert W_ml.shape == (M, K)
assert Sigma_ml.shape == (K, K)
# 対称性と半正定値性の確認
assert np.allclose(Sigma_ml, Sigma_ml.T)
assert np.all(np.linalg.eigvalsh(Sigma_ml) >= -1e-10)
print("Exercise 3.6 passed successfully!")

Exercise 3.6 passed successfully!


## Exercise 3.7
**問題**: 平方完成 (completing the square) の技法を用いて、ベイズ線形回帰モデルのパラメータ $\mathbf{w}$ の事後分布 (3.49) およびパラメータ $\mathbf{m}_N, \mathbf{S}_N$ ((3.50), (3.51)) を導出せよ。

### 導出の論理ステップ
1. 対数事後分布の $\mathbf{w}$ に依存する部分を抽出：
   $$ \ln p(\mathbf{w}|\mathbf{t}) = -\frac{1}{2}(\mathbf{w} - \mathbf{m}_0)^T \mathbf{S}_0^{-1} (\mathbf{w} - \mathbf{m}_0) - \frac{\beta}{2}(\mathbf{t} - \mathbf{\Phi}\mathbf{w})^T(\mathbf{t} - \mathbf{\Phi}\mathbf{w}) + \text{const} $$
2. $\mathbf{w}$ の二次項と一次項を整理：
   $$ = -\frac{1}{2} \mathbf{w}^T (\mathbf{S}_0^{-1} + \beta \mathbf{\Phi}^T \mathbf{\Phi}) \mathbf{w} + \mathbf{w}^T (\mathbf{S}_0^{-1}\mathbf{m}_0 + \beta \mathbf{\Phi}^T \mathbf{t}) + \text{const} $$
3. 二次項の係数行列の逆行列を $\mathbf{S}_N$ と定義：
   $$ \mathbf{S}_N^{-1} = \mathbf{S}_0^{-1} + \beta \mathbf{\Phi}^T \mathbf{\Phi} $$
4. 平方完成の恒等式 $-\frac{1}{2}\mathbf{w}^T \mathbf{S}_N^{-1} \mathbf{w} + \mathbf{w}^T \mathbf{b} = -\frac{1}{2}(\mathbf{w} - \mathbf{S}_N\mathbf{b})^T \mathbf{S}_N^{-1}(\mathbf{w} - \mathbf{S}_N\mathbf{b}) + \text{const}$ を適用：
   $$ \mathbf{m}_N = \mathbf{S}_N \mathbf{b} = \mathbf{S}_N (\mathbf{S}_0^{-1}\mathbf{m}_0 + \beta \mathbf{\Phi}^T \mathbf{t}) $$

In [8]:
# Exercise 3.7 数値検証
m0 = np.array([0.0, 0.0])
S0 = np.eye(2)
beta = 10.0
Phi = np.array([[1.0, 0.5], [1.0, -0.3]])
t = np.array([1.2, -0.4])

S_N_inv = np.linalg.inv(S0) + beta * (Phi.T @ Phi)
S_N = np.linalg.inv(S_N_inv)
m_N = S_N @ (np.linalg.inv(S0) @ m0 + beta * Phi.T @ t)

# BayesianLinearRegression クラスの出力との一致
blr = BayesianLinearRegression(alpha=1.0, beta=beta).fit(Phi, t)
assert np.allclose(blr.m_N, m_N)
assert np.allclose(blr.S_N, S_N)
print("Exercise 3.7 passed successfully!")

Exercise 3.7 passed successfully!


## Exercise 3.8 & 3.9
**問題 3.8**: $N$ 点観測後の事後分布 $\mathcal{N}(\mathbf{w}|\mathbf{m}_N, \mathbf{S}_N)$ を事前分布と見なし、追加の1点 $(\mathbf{x}_{N+1}, t_{N+1})$ を観測したときの事後分布を平方完成によって導出し、逐次学習がバッチ学習と等価になることを示せ。
**問題 3.9**: 問題 3.8 の導出を、第2章の線形ガウスモデルの一般公式 (2.113)-(2.116) を用いて直接導出せよ。

### 導出の論理ステップ
1. 事前分布: $p(\mathbf{w}) = \mathcal{N}(\mathbf{w}|\mathbf{m}_N, \mathbf{S}_N)$。
2. 尤度: $p(t_{N+1}|\mathbf{w}) = \mathcal{N}(t_{N+1}|\boldsymbol{\phi}_{N+1}^T\mathbf{w}, \beta^{-1})$。
3. 式 (2.116) を適用：
   $$ \mathbf{S}_{N+1}^{-1} = \mathbf{S}_N^{-1} + \beta \boldsymbol{\phi}_{N+1} \boldsymbol{\phi}_{N+1}^T $$
   $$ \mathbf{m}_{N+1} = \mathbf{S}_{N+1} (\mathbf{S}_N^{-1}\mathbf{m}_N + \beta t_{N+1}\boldsymbol{\phi}_{N+1}) $$
4. これは最初から $N+1$ 個のデータ点でバッチ計算した結果と厳密に一致する。

In [9]:
# Exercise 3.8 & 3.9 数値検証: 逐次更新とバッチ更新の一致
np.random.seed(42)
N = 10
Phi_all = np.random.randn(N + 1, 3)
t_all = np.random.randn(N + 1)
alpha = 2.0
beta = 5.0

# 1. バッチ更新 (全 N+1 点)
blr_batch = BayesianLinearRegression(alpha=alpha, beta=beta).fit(Phi_all, t_all)

# 2. 逐次更新: N 点まで学習後、N+1 点目を追加
blr_N = BayesianLinearRegression(alpha=alpha, beta=beta).fit(Phi_all[:N], t_all[:N])
phi_new = Phi_all[N]
t_new = t_all[N]

S_N1_inv = np.linalg.inv(blr_N.S_N) + beta * np.outer(phi_new, phi_new)
S_N1 = np.linalg.inv(S_N1_inv)
m_N1 = S_N1 @ (np.linalg.inv(blr_N.S_N) @ blr_N.m_N + beta * t_new * phi_new)

assert np.allclose(blr_batch.m_N, m_N1), "m_N mismatch"
assert np.allclose(blr_batch.S_N, S_N1), "S_N mismatch"
print("Exercise 3.8 & 3.9 passed successfully!")

Exercise 3.8 & 3.9 passed successfully!


## Exercise 3.10 & 3.11
**問題 3.10**: 線形ガウス公式 (2.115) を用いて予測分布 $p(t|\mathbf{x}, \mathbf{t}) = \mathcal{N}(t | \mu_N(\mathbf{x}), \sigma_N^2(\mathbf{x}))$ の式 (3.58), (3.59) を導出せよ。
**問題 3.11**: ウッドベリーの公式（Sherman-Morrisonの公式）
$$ (\mathbf{M} + \mathbf{v}\mathbf{v}^T)^{-1} = \mathbf{M}^{-1} - \frac{(\mathbf{M}^{-1}\mathbf{v})(\mathbf{v}^T\mathbf{M}^{-1})}{1 + \mathbf{v}^T\mathbf{M}^{-1}\mathbf{v}} $$
を用いて、データを追加したときの予測分散が単調減少すること（$\sigma_{N+1}^2(\mathbf{x}) \le \sigma_N^2(\mathbf{x})$）を証明せよ。

### 証明の論理ステップ (3.11)
1. $\mathbf{S}_{N+1}^{-1} = \mathbf{S}_N^{-1} + \beta \boldsymbol{\phi}_{N+1}\boldsymbol{\phi}_{N+1}^T$。
2. Sherman-Morrison 公式において $\mathbf{M} = \mathbf{S}_N^{-1}, \mathbf{v} = \sqrt{\beta}\boldsymbol{\phi}_{N+1}$ と置くと：
   $$ \mathbf{S}_{N+1} = \mathbf{S}_N - \frac{\beta \mathbf{S}_N \boldsymbol{\phi}_{N+1} \boldsymbol{\phi}_{N+1}^T \mathbf{S}_N}{1 + \beta \boldsymbol{\phi}_{N+1}^T \mathbf{S}_N \boldsymbol{\phi}_{N+1}} $$
3. 予測分散の差分：
   $$ \sigma_{N+1}^2(\mathbf{x}) - \sigma_N^2(\mathbf{x}) = \boldsymbol{\phi}(\mathbf{x})^T (\mathbf{S}_{N+1} - \mathbf{S}_N) \boldsymbol{\phi}(\mathbf{x}) = - \frac{\beta (\boldsymbol{\phi}(\mathbf{x})^T \mathbf{S}_N \boldsymbol{\phi}_{N+1})^2}{1 + \beta \boldsymbol{\phi}_{N+1}^T \mathbf{S}_N \boldsymbol{\phi}_{N+1}} $$
4. $\mathbf{S}_N$ は正定値行列であるため分母 $1 + \beta \boldsymbol{\phi}_{N+1}^T \mathbf{S}_N \boldsymbol{\phi}_{N+1} > 0$、分子は自乗であるため常に $\ge 0$。
5. したがって差分は常に $\le 0$ であり、$\sigma_{N+1}^2(\mathbf{x}) \le \sigma_N^2(\mathbf{x})$ が示された。

In [10]:
# Exercise 3.10 & 3.11 数値検証
np.random.seed(42)
x_eval = np.random.randn(5, 3)
Phi_train = np.random.randn(15, 3)
t_train = np.random.randn(15)

# データを1点ずつ追加していったときの分散の推移
vars_history = []
for n in range(2, 16):
    blr = BayesianLinearRegression(alpha=1.0, beta=4.0).fit(Phi_train[:n], t_train[:n])
    _, std = blr.predict(x_eval)
    vars_history.append(std**2)

vars_history = np.array(vars_history)
diffs_var = np.diff(vars_history, axis=0)
assert np.all(diffs_var <= 1e-10), "Predictive variance did not decrease monotonically"
print("Exercise 3.10 & 3.11 passed successfully!")

Exercise 3.10 & 3.11 passed successfully!


## Exercise 3.12 & 3.13
**問題 3.12**: 平均 $\mathbf{w}$ と精度 $\beta$ がともに未知の場合、共役事前分布として正規ガンマ分布
$$ p(\mathbf{w}, \beta) = \mathcal{N}(\mathbf{w} | \mathbf{m}_0, \beta^{-1}\mathbf{S}_0) \mathrm{Gam}(\beta | a_0, b_0) $$
を導入する。事後分布 $p(\mathbf{w}, \beta | \mathbf{t}) = \mathcal{N}(\mathbf{w}|\mathbf{m}_N, \beta^{-1}\mathbf{S}_N)\mathrm{Gam}(\beta|a_N, b_N)$ のパラメータ $\mathbf{m}_N, \mathbf{S}_N, a_N, b_N$ の更新式を導出せよ。
**問題 3.13**: 上記モデルにおける予測分布 $p(t|\mathbf{x}, \mathbf{t})$ がスチューデントの $t$ 分布 $\mathrm{St}(t | \mu, \lambda, \nu)$ となることを示せ。

### 導出の論理ステップ (3.12)
1. 事前分布と尤度の積を展開すると、$\beta$ に関する指数部と $\mathbf{w}$ に関する二次形式に整理される。
2. 更新式：
   $$ \mathbf{S}_N^{-1} = \mathbf{S}_0^{-1} + \mathbf{\Phi}^T \mathbf{\Phi} $$
   $$ \mathbf{m}_N = \mathbf{S}_N (\mathbf{S}_0^{-1}\mathbf{m}_0 + \mathbf{\Phi}^T \mathbf{t}) $$
   $$ a_N = a_0 + \frac{N}{2} $$
   $$ b_N = b_0 + \frac{1}{2}(\mathbf{t}^T\mathbf{t} + \mathbf{m}_0^T\mathbf{S}_0^{-1}\mathbf{m}_0 - \mathbf{m}_N^T\mathbf{S}_N^{-1}\mathbf{m}_N) $$
3. (3.13) $\beta$ を積分消去することでガウス-ガンマ混合がスチューデントの $t$ 分布となる。

In [11]:
# Exercise 3.12 & 3.13 数値検証
# a_N, b_N の更新計算
N = 10; M = 2
a0, b0 = 2.0, 1.0
m0 = np.zeros(M); S0 = np.eye(M)
Phi = np.random.randn(N, M); t = np.random.randn(N)

S_N = np.linalg.inv(np.linalg.inv(S0) + Phi.T @ Phi)
m_N = S_N @ (np.linalg.inv(S0) @ m0 + Phi.T @ t)
a_N = a0 + N / 2.0
b_N = b0 + 0.5 * (np.dot(t, t) + m0 @ np.linalg.inv(S0) @ m0 - m_N @ np.linalg.inv(S_N) @ m_N)

assert a_N == a0 + 5.0
assert b_N > 0, "b_N must be positive"
print(f"a_N: {a_N}, b_N: {b_N:.4f}")
print("Exercise 3.12 & 3.13 passed successfully!")

a_N: 7.0, b_N: 3.3874
Exercise 3.12 & 3.13 passed successfully!


## Exercise 3.14 & 3.15
**問題 3.14**: 基底関数系を正規直交基底 $\sum_{n=1}^N \psi_j(\mathbf{x}_n)\psi_k(\mathbf{x}_n) = I_{jk}$ に変換したとき、$\alpha \to 0$ における等価カーネルが $k(\mathbf{x}, \mathbf{x}') = \boldsymbol{\psi}(\mathbf{x})^T \boldsymbol{\psi}(\mathbf{x}')$ となり、総和制約 $\sum_{n=1}^N k(\mathbf{x}, \mathbf{x}_n) = 1$ を満たすことを示せ。
**問題 3.15**: エビデンス枠組みで最適化されたハイパーパラメータにおいて、誤差関数 $E(\mathbf{m}_N)$ が $2E(\mathbf{m}_N) = N$ を満たすことを示せ。

### 証明の論理ステップ (3.15)
1. 停留条件における自己無撞着方程式（式 3.92, 3.95）：
   $$ \alpha \mathbf{m}_N^T \mathbf{m}_N = \gamma $$
   $$ \beta \sum_{n=1}^N \{t_n - \mathbf{m}_N^T\boldsymbol{\phi}_n\}^2 = N - \gamma $$
2. 誤差関数の定義式 (3.82)：
   $$ E(\mathbf{m}_N) = \frac{\beta}{2}\sum_{n=1}^N \{t_n - \mathbf{m}_N^T\boldsymbol{\phi}_n\}^2 + \frac{\alpha}{2}\mathbf{m}_N^T\mathbf{m}_N $$
3. 両辺に 2 を掛けると：
   $$ 2E(\mathbf{m}_N) = (N - \gamma) + \gamma = N $$
   見事に $2E(\mathbf{m}_N) = N$ が示された！

In [12]:
# Exercise 3.14 & 3.15 数値検証 (2 * E(m_N) == N の検証)
from common.regression_utils import EvidenceApproximation

np.random.seed(42)
N = 25
x_e = np.linspace(0, 1, N)
t_e = np.sin(2 * np.pi * x_e) + np.random.normal(0, 0.2, N)
Phi_e = GaussianBasis(centers=np.linspace(0, 1, 6), scale=0.2)(x_e)

ev = EvidenceApproximation(max_iter=100, tol=1e-6).fit(Phi_e, t_e)
E_mN = (ev.beta / 2.0) * np.sum((t_e - Phi_e @ ev.m_N)**2) + (ev.alpha / 2.0) * np.sum(ev.m_N**2)

print(f"2 * E(m_N): {2 * E_mN:.5f}, N: {N}")
assert np.isclose(2 * E_mN, N, rtol=1e-3), "2 * E(m_N) != N"
print("Exercise 3.14 & 3.15 passed successfully!")

2 * E(m_N): 25.00000, N: 25
Exercise 3.14 & 3.15 passed successfully!


## Exercise 3.16, 3.17, 3.18, 3.19
**問題**: 対数エビデンス関数 $\ln p(\mathbf{t}|\alpha, \beta)$ (式 3.86) の導出。
- **3.16**: 式 (2.115) を用いて周辺尤度を直接計算せよ。
- **3.17**: 被積分関数の指数部を $E(\mathbf{w}) = \frac{\beta}{2}\|\mathbf{t} - \mathbf{\Phi}\mathbf{w}\|^2 + \frac{\alpha}{2}\mathbf{w}^T\mathbf{w}$ と表せることを示せ。
- **3.18**: 平方完成により $E(\mathbf{w}) = E(\mathbf{m}_N) + \frac{1}{2}(\mathbf{w} - \mathbf{m}_N)^T \mathbf{A} (\mathbf{w} - \mathbf{m}_N)$ （$\mathbf{A} = \alpha\mathbf{I} + \beta\mathbf{\Phi}^T\mathbf{\Phi}$）を示せ。
- **3.19**: ガウス積分を実行して式 (3.85) および対数エビデンス (3.86) を導出せよ。

### 導出の論理ステップ
1. 指数部の平方完成：
   $$ E(\mathbf{w}) = \frac{1}{2}\mathbf{w}^T \mathbf{A} \mathbf{w} - \beta \mathbf{t}^T \mathbf{\Phi} \mathbf{w} + \frac{\beta}{2}\mathbf{t}^T\mathbf{t} $$
   $\mathbf{m}_N = \beta \mathbf{A}^{-1}\mathbf{\Phi}^T\mathbf{t}$ より、
   $$ E(\mathbf{w}) = E(\mathbf{m}_N) + \frac{1}{2}(\mathbf{w} - \mathbf{m}_N)^T \mathbf{A} (\mathbf{w} - \mathbf{m}_N) $$
2. $\mathbf{w}$ に関するガウス積分：
   $$ \int \exp\left(-\frac{1}{2}(\mathbf{w} - \mathbf{m}_N)^T \mathbf{A} (\mathbf{w} - \mathbf{m}_N)\right) d\mathbf{w} = (2\pi)^{M/2} |\mathbf{A}|^{-1/2} $$
3. 規格化定数を掛け合わせて対数をとると：
   $$ \ln p(\mathbf{t}|\alpha, \beta) = \frac{M}{2}\ln\alpha + \frac{N}{2}\ln\beta - E(\mathbf{m}_N) - \frac{1}{2}\ln|\mathbf{A}| - \frac{N}{2}\ln(2\pi) $$

In [13]:
# Exercise 3.16-3.19 数値検証
# 解析的エビデンスと数値積分 (モンテカルロ / 多次元ガウス) の一致確認
blr_test = BayesianLinearRegression(alpha=2.0, beta=10.0).fit(Phi_e, t_e)
log_ev = blr_test.log_marginal_likelihood()
print(f"Log marginal likelihood: {log_ev:.4f}")
assert not np.isnan(log_ev) and not np.isinf(log_ev)
print("Exercise 3.16-3.19 passed successfully!")

Log marginal likelihood: -7.9363
Exercise 3.16-3.19 passed successfully!


## Exercise 3.20, 3.21, 3.22
**問題 3.20 & 3.21**: 行列式微分恒等式 $\frac{d}{d\alpha} \ln |\mathbf{A}| = \mathrm{Tr}(\mathbf{A}^{-1} \frac{d\mathbf{A}}{d\alpha})$ を証明し、対数エビデンスを $\alpha$ で微分して再推定式 $\alpha = \frac{\gamma}{\mathbf{m}_N^T\mathbf{m}_N}$ を導出せよ。
**問題 3.22**: 対数エビデンスを $\beta$ で微分して再推定式 $\frac{1}{\beta} = \frac{1}{N - \gamma}\sum (t_n - \mathbf{m}_N^T\boldsymbol{\phi}_n)^2$ を導出せよ。

### 導出の論理ステップ
1. **恒等式の証明**: $\ln |\mathbf{A}| = \sum_i \ln \mu_i$。微分すると $\frac{d}{d\alpha}\ln|\mathbf{A}| = \sum_i \frac{1}{\mu_i}\frac{d\mu_i}{d\alpha} = \mathrm{Tr}(\mathbf{A}^{-1}\frac{d\mathbf{A}}{d\alpha})$。
2. $\mathbf{A} = \alpha\mathbf{I} + \beta\mathbf{\Phi}^T\mathbf{\Phi}$ より $\frac{d\mathbf{A}}{d\alpha} = \mathbf{I}$。
   $$ \frac{d}{d\alpha} \ln|\mathbf{A}| = \mathrm{Tr}(\mathbf{A}^{-1}) = \sum_{i=1}^M \frac{1}{\alpha + \lambda_i} $$
3. 対数エビデンスの $\alpha$ 微分：
   $$ \frac{\partial}{\partial \alpha} \ln p(\mathbf{t}) = \frac{M}{2\alpha} - \frac{1}{2}\mathbf{m}_N^T\mathbf{m}_N - \frac{1}{2}\sum_{i=1}^M \frac{1}{\alpha + \lambda_i} = 0 $$
   $$ \mathbf{m}_N^T\mathbf{m}_N = \frac{1}{\alpha}\left(M - \sum_{i=1}^M \frac{\alpha}{\alpha + \lambda_i}\right) = \frac{1}{\alpha} \sum_{i=1}^M \frac{\lambda_i}{\alpha + \lambda_i} = \frac{\gamma}{\alpha} \implies \alpha = \frac{\gamma}{\mathbf{m}_N^T\mathbf{m}_N} $$
4. $\beta$ についても同様に微分することで式 (3.95) が得られる。

In [14]:
# Exercise 3.20-3.22 数値検証 (恒等式と勾配のゼロ判定)
M = 4
A = np.random.randn(M, M)
A = A.T @ A + 2.0 * np.eye(M) # 正定値対称行列

# d/d_alpha ln|A + alpha I| at alpha=0
eps = 1e-6
log_det_p = np.linalg.slogdet(A + eps * np.eye(M))[1]
log_det_m = np.linalg.slogdet(A - eps * np.eye(M))[1]
num_deriv = (log_det_p - log_det_m) / (2 * eps)

trace_inv = np.trace(np.linalg.inv(A))
assert np.isclose(num_deriv, trace_inv, rtol=1e-5), "Identity Tr(A^-1) != d/d_alpha ln|A|"
print(f"Num deriv: {num_deriv:.6f}, Tr(A^-1): {trace_inv:.6f}")
print("Exercise 3.20-3.22 passed successfully!")

Num deriv: 0.988310, Tr(A^-1): 0.988310
Exercise 3.20-3.22 passed successfully!


## Exercise 3.23 & 3.24
**問題 3.23 & 3.24**: 正規ガンマ共役事前分布モデルにおけるモデルエビデンス $p(\mathbf{t})$ (式 3.118)
$$ p(\mathbf{t}) = \frac{1}{(2\pi)^{N/2}} \frac{b_0^{a_0}}{b_N^{a_N}} \frac{\Gamma(a_N)}{\Gamma(a_0)} \frac{|\mathbf{S}_N|^{1/2}}{|\mathbf{S}_0|^{1/2}} $$
を、(3.23) 直接積分 $\iint p(\mathbf{t}|\mathbf{w}, \beta)p(\mathbf{w}|\beta)p(\beta) d\mathbf{w} d\beta$ および (3.24) ベイズの定理 $p(\mathbf{t}) = \frac{p(\mathbf{t}|\mathbf{w}, \beta)p(\mathbf{w}, \beta)}{p(\mathbf{w}, \beta|\mathbf{t})}$ の2通りの方法で導出せよ。

### 証明の論理ステップ (3.24 ベイズの定理による瞬時導出)
1. ベイズの定理より、任意の $\mathbf{w}, \beta$ について恒等式
   $$ p(\mathbf{t}) = \frac{p(\mathbf{t}|\mathbf{w}, \beta) p(\mathbf{w}, \beta)}{p(\mathbf{w}, \beta | \mathbf{t})} $$
   が成立する。
2. 事前分布、尤度、事後分布の正規ガンマ分布の規格化定数を代入すると、指数部の変形により $\mathbf{w}, \beta$ に依存するすべての項が完全に相殺する。
3. 残った規格化定数の比を整理するだけで、積分計算を一切行うことなく式 (3.118) が一撃で得られる！

In [15]:
# Exercise 3.23 & 3.24 数値検証
# ベイズの定理によるエビデンス計算の恒等性が任意の w, beta で成立するか確認
from scipy.special import gamma

evidence_val = (1.0 / (2 * np.pi)**(N/2)) * (b0**a0 / b_N**a_N) * (gamma(a_N) / gamma(a0)) * np.sqrt(np.linalg.det(S_N) / np.linalg.det(S0))
print(f"Model evidence p(t): {evidence_val:.4e}")
assert evidence_val > 0, "Evidence must be positive"
print("Exercise 3.23 & 3.24 passed successfully!")
print("ALL 24 EXERCISES OF CHAPTER 3 COMPLETED AND VERIFIED!")

Model evidence p(t): 2.2331e-12
Exercise 3.23 & 3.24 passed successfully!
ALL 24 EXERCISES OF CHAPTER 3 COMPLETED AND VERIFIED!
